src/main/java/com/cloudbees/hudson/plugins/folder/computed/EventOutputStreams.java

```diff
@@ -0,0 +1,270 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.cloudbees.hudson.plugins.folder.computed;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.io.Closeable;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Arrays;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.commons.io.IOUtils;
+
+/**
+ * An factory for {@link OutputStream} instances that can concurrently write to the same file and do lots of other
+ * wonderful magical things (excluding changing the baby's nappy).
+ * <ul>
+ * <li>Output is collected in batches, one batch per thread and flushed periodically.</li>
+ * <li>The file is only opened when actually writing.</li>
+ * <li>The file is rotated if it gets too big.</li>
+ * <li>You can close it over and over, it will auto re-open on next write!</li>
+ * </ul>
+ * The primary intended use case is for event processing with {@link FolderComputation#createEventsListener()}
+ * where there may be multiple concurrent events in flight and we need to:
+ * <ul>
+ * <li>Try to keep each events log messages close together, hence the batching</li>
+ * <li>Rotate after it gets too big</li>
+ * <li>Not hold the file open indefinitely - ideally only when writing so that it can be moved by other processes</li>
+ * </ul>
+ * @since 5.18
+ */
+public class EventOutputStreams implements Closeable {
+    /**
+     * The queue of pending output.
+     */
+    private final BlockingQueue<byte[]> pending = new LinkedBlockingQueue<byte[]>();
+    /**
+     * The amount of data in the pending queue.
+     */
+    private final AtomicInteger pendingSize = new AtomicInteger();
+    /**
+     * The primary file name.
+     */
+    private final File primaryFile;
+    /**
+     * How often to trigger a flush (defaults to 250ms)
+     */
+    private final long flushIntervalNanos;
+    /**
+     * How big to trigger a flush (defaults to 1024)
+     */
+    private final int flushSize;
+    /**
+     * How big to trigger a rotation (defaults to 32k)
+     */
+    private final long rotateSize;
```

Comment:
The version in core will roll on startup and has other issues too, so partial reuse but necessary. I may duplicate this class in core and switch when available